code-dot-org/code-dot-org
# Embed CSF deepdive Jotform surveys to our site

**What**
1. Add new controller actions to handle CSF survey requests
  - Pre-workshop survey link `http://studio.code.org/pd/workshop_survey/csf/pre201`.\
Only for teachers enroll in an CSF 201 workshop that has not ended.
  - Post-workshop survey links `http://studio.code.org/pd/workshop_survey/csf/post201`\
or `http://studio.code.org/pd/workshop_survey/csf/post201/<10-char-enrollment-code>`.\
Only for teachers attend an CSF 201 workshop.
  - Send `userId, username, userEmail` to Jotform forms.

2. Require log-in when enrolling in CSF Deep dive

**Why**
The main reasons to create a separate controller action for CSF instead of re-using `new_general` (for pre and daily CSD/CSP workshop surveys) or `new_post` (for post CSD/CSP workshop survey) are
1. Have a more descriptive URL `/pd/workshop_survey/csf/(pre201|post201)` instead of `/pd/workshop_survey/day/0|1`
2. Allow 1 post-workshop URL for everyone `/pd/workshop_survey/csf/post201` (in which enrollment code is optional) instead of `/pd/workshop_survey/post/<enrollment-code>`
3. Avoid regression in CSD/CSP actions.
4. Easier to add more surveys in the future e.g. `/pd/workshop_surve/csf/(pre301|post301|csf/daily301/1)`

**How tested**

- Manual test:
  - Create a test Jotform form and get its id (e.g. 90845160121145)
  - Add Jotform ids hash to `locals.yml`, it looks like this
    ```
    jotform_forms:
      csf:
        pre201: 90845160121145
        post201: 90845160121145
    ```
  - In local ruby console, create 2 CSF deepdive workshops and enroll your teacher account in those workshops. Attend one of them.
    ```
    include FactoryGirl::Syntax::Methods
    u = User.find_by_email('ha@code.org')
    ws_not_started = create :pd_workshop, course: Pd::Workshop::COURSE_CSF, subject: Pd::Workshop::SUBJECT_CSF_201, num_sessions: 1, sessions_from: DateTime.now + 1.days
    _ = create :pd_enrollment, :from_user, user: u, workshop: ws_not_started
    Pd::Workshop.enrolled_in_by(u).nearest   #this should return ws_not_started

    ws_started = create :pd_workshop, course: Pd::Workshop::COURSE_CSF, subject: Pd::Workshop::SUBJECT_CSF_201, num_sessions: 1, started_at: DateTime.now - 7.days
    enroll = create :pd_enrollment, :from_user, user: u, workshop: ws_started
    enroll.code
    attend = create :pd_attendance, session: ws_started.sessions[0], teacher: u
    Pd::Workshop.with_nearest_attendance_by(u)  # this should return ws_started
    ``` 
  - Access pre-deepdive workshop from `http://localhost-studio.code.org:3000/pd/workshop_survey/csf/pre201/`
  - Access post-deepdive workshop from `http://localhost-studio.code.org:3000/pd/workshop_survey/csf/post201/` and `http://localhost-studio.code.org:3000/pd/workshop_survey/csf/post201/<enroll.code>`
  - Go all the way to survey submission.
  - To try again after submissions, delete existing survey in local db by running `Pd::WorkshopDailySurvey.destroy_all` in ruby console

- (Automated) Integration test:
In dashboard, runs `bin/rails test test/controllers/pd/workshop_daily_survey_controller_test.rb -n /csf/` to execute 7 tests for pre-deepdive survey and 9 for post-deepdive survey.
  - csf pre201 survey: unauthenticated teacher is redirected to sign-in
  - csf pre201 survey: unenrolled teacher gets not_enrolled msg
  - csf pre201 survey: enrolled teacher in ended workshop gets too-late msg
  - csf pre201 survey: enrolled teacher in unended workshop gets survey
  - csf pre201 survey: reports survey render to New Relic
  - csf pre201 survey: submission creates a placeholder record and redirects teacher to thanks
  - csf pre201 survey: teacher already submitted survey does not gets survey again
  - csf post201 survey: unauthenticated teacher is redirected to sign-in
  - csf post201 survey: unenrolled teacher gets not-enrolled msg
  - csf post201 survey: unattended teacher gets no-attendance msg
  - csf post201 survey: attended teacher with invalid enrollment code gets 404 msg
  - csf post201 survey: attended teacher with valid enrollment code gets survey
  - csf post201 survey: attended teacher without enrollment code gets survey
  - csf post201 survey: reports survey render to New Relic
  - csf post201 survey: submission creates a placeholder record and redirects teacher to thanks
  - csf post201 survey: teacher already submitted survey does not gets survey again

Also `bin/rails test test/controllers/pd/workshop_enrollment_controller_test.rb` for the change in workshop_enrollment_controller.rb
